ddtrace/profiling/collector/stack.pyx

```diff
@@ -272,94 +303,81 @@ cdef stack_collect(ignore_profiler, thread_time, max_nframes, interval, wall_tim
-                                (tstate.thread_id, <object>exc_info.exc_type, <object>exc_info.exc_traceback)
-                            )
+                            current_exceptions[tstate.thread_id] = (<object>exc_info.exc_type, <object>exc_info.exc_traceback)
 
                         tstate = PyThreadState_Next(tstate)
 
                     interp = PyInterpreterState_Next(interp)
             finally:
                 PyThread_release_lock(lmutex)
     ELSE:
-        running_threads = list(sys._current_frames().items())
-
-    running_thread_ids = {t[0] for t in running_threads}
-
-    if thread_span_links:
-        thread_span_links.clear_threads(running_thread_ids)
+        cdef extern from "<pystate.h>":
+            PyObject* _PyThread_CurrentFrames()
+
+        cdef dict running_threads = <dict>_PyThread_CurrentFrames()
+
+    cdef dict cpu_times = thread_time(running_threads.keys())
+
+    return tuple(
+        (
+            pthread_id,
+            native_thread_id,
+            get_thread_name(pthread_id),
+            running_threads[pthread_id],
+            current_exceptions.get(pthread_id),
+            thread_span_links.get_active_leaf_spans_from_thread_id(pthread_id) if thread_span_links else set(),
+            cpu_time,
+        )
+        for (pthread_id, native_thread_id), cpu_time in cpu_times.items()
+        if not ignore_profiler or pthread_id not in _periodic.PERIODIC_THREAD_IDS
+    )
 
-    if ignore_profiler:
-        running_thread_ids -= _periodic.PERIODIC_THREAD_IDS
 
-    # Build a dict of pthread_t id -> native thread id
-    # We need to use both as pthread_id can be reused for new thread, messing with the CPU time clock
-    thread_native_ids = {}
 
-    for thread_id in running_thread_ids:
-        try:
-            thread_obj = threading._active[thread_id]
-        except KeyError:
-            # This should not happen, unless somebody started a thread without
-            # using the `threading` module; in that case, ignore the thread altogether.
-            # In that case, well… just use the thread_id as native_id 🤞
-            native_id = thread_id
-        else:
-            # We prioritize using native ids since we expect them to be surely unique for a program. This is less true
-            # for hashes since they are relative to the memory address which can easily be the same across different
-            # objects.
-            try:
-                native_id = thread_obj.native_id
-            except AttributeError:
-                # Python < 3.8
-                native_id = hash(thread_obj)
+cdef stack_collect(ignore_profiler, thread_time, max_nframes, interval, wall_time, thread_span_links):
 
-        thread_native_ids[thread_id] = native_id
+    running_threads = collect_threads(ignore_profiler, thread_time, thread_span_links)
 
-    cpu_time = thread_time(thread_native_ids)
+    if thread_span_links:
+        # FIXME also use native thread id
```

Comment:
Can this be done now?

Reply:
Not more now than before (the `FIXME` was already there).
It'll wait for another PR. :)